File: packages/lynx-background/android/com/sigx/background/BackgroundEventBus.kt
package com.sigx.background

import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.react.bridge.JavaOnlyMap
import com.lynx.tasm.LynxView
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap

/**
 * Process-wide pub/sub bus that converts WorkManager fires into JS-side
 * event payloads, and routes JS completions back to the suspended worker so
 * `doWork()` can return `Result.success` / `Result.failure`.
 *
 * Mirrors [com.sigx.notifications.PushEventBus]. Decoupled from any specific
 * LynxView so workers fired before a view exists can still publish — the
 * worker side gates on the completion future, with a bounded grace period
 * to wait for a JS handler to bind.
 *
 * Wire shape on the `__sigxBackgroundFire` channel:
 *     { "taskName": String, "runId": String }
 */
internal object BackgroundEventBus {

    /**
     * Single lock guards [listeners]. Listener invocations happen OUTSIDE
     * the lock so a callback that re-enters the bus doesn't deadlock.
     */
    private val lock = Any()
    private val listeners = mutableListOf<Pair<UUID, (String, JavaOnlyMap) -> Unit>>()

    /**
     * Completion futures keyed by `runId`. The worker awaits these; JS
     * calls [completeRun] from `Background.completeTask(runId, success)`.
     *
     * `ConcurrentHashMap` so the bus + worker thread + JS bridge thread can
     * all touch it without taking the listener lock.
     */
    private val inflight = ConcurrentHashMap<String, CompletableFuture<Boolean>>()

    fun addListener(fn: (String, JavaOnlyMap) -> Unit): UUID {
        val token = UUID.randomUUID()
        synchronized(lock) { listeners.add(token to fn) }
        return token
    }

    fun removeListener(token: UUID) {
        synchronized(lock) { listeners.removeAll { it.first == token } }
    }

    /**
     * Called by [SigxBackgroundWorker]. Registers a completion future and
     * publishes a fire event to JS. The caller awaits the returned future
     * (with a timeout) to decide the worker's `Result`.
     */
    fun publishFire(taskName: String, runId: String): CompletableFuture<Boolean> {
        val future = CompletableFuture<Boolean>()
        inflight[runId] = future
        val payload = JavaOnlyMap().apply {
            putString("taskName", taskName)
            putString("runId", runId)
        }
        emit(Channel.FIRE, payload)
        return future
    }

    /** Called from `Background.completeTask`. No-op if the run has already timed out. */
    fun completeRun(runId: String, success: Boolean) {
        inflight.remove(runId)?.complete(success)
    }

    /** Called by the worker if it times out — releases the future entry. */
    fun abandonRun(runId: String) {
        inflight.remove(runId)
    }

    private fun emit(channel: String, payload: JavaOnlyMap) {
        // Snapshot inside the lock so a concurrent add/remove can't tear the
        // iteration. Call listeners OUTSIDE so a callback that re-enters
        // doesn't deadlock.
        val snapshot = synchronized(lock) { listeners.toList() }
        for ((_, fn) in snapshot) fn(channel, payload)
    }

    object Channel {
        const val FIRE = "__sigxBackgroundFire"
    }
}

File: packages/lynx-background/ios/BackgroundEventBus.swift
import Foundation

/// Process-wide pub/sub bus that converts BGTaskScheduler fire callbacks into
/// JS-side event payloads, and routes JS completions back to the in-flight
/// `BGTask` so the OS can hear `setTaskCompleted(success:)`.
///
/// Mirrors `PushEventBus` in `@sigx/lynx-notifications`. Owned process-wide,
/// independent of any LynxView lifecycle — BGTasks can fire before a LynxView
/// exists, and the bus replays nothing (each fire is a new OS event with its
/// own deadline; replaying a stale runId would race the OS expiration).
///
/// Wire shape on the `__sigxBackgroundFire` channel:
///     { "taskName": String, "runId": String }
final class BackgroundEventBus {

    static let shared = BackgroundEventBus()

    typealias Payload = [String: Any]
    typealias Listener = (String, Payload) -> Void

    private let queue = DispatchQueue(label: "com.sigx.background.bus")
    private var listeners: [(token: UUID, fn: Listener)] = []

    /// In-flight BGTasks, keyed by the `runId` we hand to JS. When JS calls
    /// `Background.completeTask(runId, success)` we look up the BGTask here
    /// and call `setTaskCompleted(success:)` on it. Stored as `Any` because
    /// `BGTask` is only available on iOS 13+ and this file builds against
    /// older deployment targets too — the cast site checks availability.
    private var inflight: [String: Any] = [:]

    // MARK: - Listener management

    @discardableResult
    func addListener(_ fn: @escaping Listener) -> UUID {
        let token = UUID()
        queue.sync { listeners.append((token, fn)) }
        return token
    }

    func removeListener(_ token: UUID) {
        queue.sync { listeners.removeAll { $0.token == token } }
    }

    // MARK: - Fire / complete

    /// Called by `BackgroundScheduler` from inside the BGTaskScheduler launch
    /// handler. Records the task under a fresh `runId`, publishes the fire
    /// event to JS, and returns the `runId` so the caller can wire the OS
    /// expiration handler to fail the same `runId`.
    func publishFire(taskName: String, task: Any) -> String {
        let runId = UUID().uuidString
        queue.sync { inflight[runId] = task }
        emit(
            channel: BackgroundEventChannel.fire,
            payload: ["taskName": taskName, "runId": runId],
        )
        return runId
    }

    /// Drain and return the in-flight task for a `runId`. Returns nil if the
    /// `runId` has already been completed (double-completion is harmless —
    /// JS may race the OS expiration handler).
    func takeInflight(runId: String) -> Any? {
        return queue.sync {
            let v = inflight[runId]
            inflight[runId] = nil
            return v
        }
    }

    private func emit(channel: String, payload: Payload) {
        let snapshot = queue.sync { listeners }
        for (_, fn) in snapshot { fn(channel, payload) }
    }
}

/// Stable event-channel names. JS shim mirrors these in `src/events.ts` —
/// keep in sync.
enum BackgroundEventChannel {
    static let fire = "__sigxBackgroundFire"
}
